sms-authenticator/src/main/java/netzbegruenung/keycloak/authenticator/PhoneNumberRequiredAction.java

```diff
@@ -129,10 +131,52 @@ public void evaluateTriggers(RequiredActionContext context) {
 		}
 	}
 
+	/**
+	 * Generates a list of country codes with their names and emojis to be displayed in the phone number input form.
+	 *
+	 * @param context the current RequiredActionContext
+	 * @return a list of maps containing country name, code, and emoji
+	 */
+	public List<Map<String, String>> getCountryCodeList(RequiredActionContext context) {
+		AuthenticatorConfigModel config = context.getRealm().getAuthenticatorConfigByAlias("sms-2fa");
+		String countryCodeList = config.getConfig().getOrDefault("countryCodeList", "");
+
+		UserModel user = context.getUser();
+		KeycloakSession session = context.getSession();
+		Locale locale = session.getContext().resolveLocale(user);
+
+		PhoneNumberUtil phoneUtil = PhoneNumberUtil.getInstance();
+
+		List<Map<String, String>> countryList = new ArrayList<>();
+		if (!countryCodeList.isBlank()) {
+			List<String> countryCodes = Arrays.asList(countryCodeList.split(","));
+			for (String countryCode : countryCodes) {
+				try {
+					String code = "+" + Integer.toString(phoneUtil.getCountryCodeForRegion(countryCode.trim().toUpperCase()));
+					String countryName = new Locale("", countryCode.trim().toUpperCase()).getDisplayCountry(locale);
+
+					// generate emoji from country code
+					int offset = 0x1F1E6;  // Base Unicode for regional indicator symbols
+					int codePoint1 = offset + (countryCode.trim().toUpperCase().charAt(0) - 'A');
+					int codePoint2 = offset + (countryCode.trim().toUpperCase().charAt(1) - 'A');
+					String emoji = new String(new int[]{codePoint1, codePoint2}, 0, 2);
+
+					countryList.add(Map.of("name", countryName, "code", code, "emoji", emoji));
+					logger.infof("Added country code %s for country %s", code, countryName);
```

Comment:
You can make the loop slightly cleaner by doing the trim().toUpperCase() operation once per item. You could also use a Java Stream to make the logic more declarative.

The emoji generation is a neat piece of logic. You could extract it into its own private helper method to make the main getCountryCodeList method easier to read.

```
countryCodes.stream()
        .map(String::trim)
        .map(String::toUpperCase)
        .map(code -> createCountryInfo(code, userLocale, phoneUtil))
        .filter(java.util.Objects::nonNull) // Filter out any that failed to process
        .collect(Collectors.toList());
```